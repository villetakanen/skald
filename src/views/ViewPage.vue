<template>
  <div :class="theme">
    <div class="page-container">
      <v-container v-if="loading">
        <v-row>
          <v-col align='center'>
            <Loading/>
          </v-col>
        </v-row>
      </v-container>
      <v-container v-if="!loading">
        <h1 class="sitelink">{{siteid}}</h1>
        <h1 class="pagetitle">{{title}}</h1>

        <v-row>

        <v-col
            cols="12"
            md="8">
          <div id="reader-text">
            <v-card>
              <v-container>
                <WikiText :content="content" :siteid="siteid"/>
              </v-container>
            <v-speed-dial
                v-model="fab"
                top
                right
                direction="bottom"
                absolute
                style="margin-top:-32px">
                <template v-slot:activator>
                  <v-btn
                    v-model="fab"
                    color="secondary"
                    small
                    top
                    right
                    fab
                    elevation="2">
                    <v-icon v-if="fab">mdi-close</v-icon>
                    <v-icon v-else>mdi-menu-down</v-icon>
                  </v-btn>
                </template>
                 <v-btn
                  :disabled="!isAuthz"
                  fab
                  small
                  color="primary"
                  v-bind:to="editlink"
                  >
                  <v-icon>mdi-pencil</v-icon>
                </v-btn>
                <v-btn
                  fab
                  dark
                  small
                  color="primary"
                  @click="copyUrl"
                >
                  <v-icon>mdi-link</v-icon>
                </v-btn>
              </v-speed-dial>
          </v-card>
          </div>
          </v-col>

          <v-col
            cols="12"
            md="4">
            <v-card>
              <v-card-text>
                <div class="sidebar">
                  <WikiText :content="sidebarContent" :siteid="siteid"/>
                </div>
            </v-card-text>
              <v-card-actions v-if="isAuthz">
                <v-btn
                  text
                  color="primary"
                  :to="`/e/${siteid}/sidebar`">Edit sidebar</v-btn>
              </v-card-actions>
            </v-card>
          </v-col>
          <v-col xs12>
          <v-card>
            <v-card-text>
              <LatestChanges />
            </v-card-text>
          </v-card>
          </v-col>
        </v-row>
      </v-container>
    </div>
  </div>
</template>

<script>
import LatestChanges from '../components/LatestChanges'
import WikiText from '../components/WikiText'
import Loading from '../components/Loading'

export default {
  components: {
    LatestChanges,
    WikiText,
    Loading
  },
  props: [
    'pageid',
    'siteid'
  ],
  data: () => ({
    fab: false
  }),
  created () {
    this.updatePage(this.siteid, this.pageid)
  },
  watch: {
    '$route' (to, from) {
      this.updatePage(this.siteid, this.pageid)
    }
  },
  computed: {
    isAuthz () {
      return this.$store.getters['isAuthz']()
    },
    title () {
      if (this.$store.getters['binder/title']() === null) return ' '
      return this.$store.getters['binder/title']()
    },
    content () {
      if (this.$store.getters['binder/content']() === null) return ' '
      return this.$store.getters['binder/content']()
    },
    sidebarContent () {
      if (this.$store.getters['site/sidebarContent']() === null) return ' '
      return this.$store.getters['site/sidebarContent']()
    },
    theme () {
      if (this.$store.getters['binder/theme']() === null) return 'Skald reader'
      return this.$store.getters['binder/theme']() + ' reader'
    },
    loading () {
      return this.$store.getters['binder/loading']()
    },
    editlink () {
      // If we are showing the root of site, no pageid is set.
      if (!this.pageid) return `/e/${this.siteid}/${this.siteid}`
      return `/e/${this.siteid}/${this.pageid}`
    }
  },
  methods: {
    updatePage (siteid, pageid) {
      if (siteid === null || typeof siteid === 'undefined') siteid = 'skald'
      if (pageid === null || typeof pageid === 'undefined') pageid = siteid
      this.$store.dispatch('binder/openPage', { siteid: siteid, pageid: pageid })

      // set site styles
      this.$store.dispatch('sites/openSite', siteid)
      this.$store.dispatch('site/open', siteid)

      // in case we are updating the same page, without route change:
      // we need to manually scroll to top
      window.scroll(0, 0)
    },
    copyUrl () {
      const dummy = document.createElement('input')
      document.body.appendChild(dummy)
      dummy.value = window.location.href
      dummy.select()
      document.execCommand('copy')
      document.body.removeChild(dummy)
      this.$store.commit('snack', 'Url copied to clipboard')
    }
  }
}
</script>

<style lang="scss">
h1.sitelink,
h1.pagetitle {
  text-shadow: 0px 0px 8px rgba(0,0,0,1);
  font-weight: 500;
}
h1.sitelink{
  font-size: 14px;
  line-height: 16px;
}
h1.pagetitle{
  font-size:28px;
}
div.infodocpart{
  border: 1px solid #666;
  padding: 8px;
  border-radius: 4px;
  background-color: rgba(0,0,0,0.2)
}
</style>
