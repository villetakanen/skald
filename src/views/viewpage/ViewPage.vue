<template>
  <div class="page-container">
    <v-container >
      <v-row v-if="meta.loading">
        <v-col>
          <Loading
            center
            class="ma-4"/>
        </v-col>
      </v-row>
      <v-row>
        <v-col>
          <TabTitle
            :sub="site.name"
            :sublink="`/v/${site.siteid}`"
            :topic="page.name"
            :link="`/v/${site.siteid}/${page.pageid}`"/>
        </v-col>
      </v-row>
      <v-row v-if="!meta.loading">
        <v-col cols="12" md="8">
          <v-card>
            <v-card-text>
                <div class="wikipage">
                    <div :class="site.theme">
                      <WikiContent :html="content"/>
                    </div>
                </div>
                <v-btn
                  :disabled="!editorActions"
                  fab
                  small
                  top right absolute
                  style= "margin-right: 48px; margin-top:4px"
                  color="primary"
                  v-bind:to="`/edit/${site.siteid}/${page.pageid}`"
                  ><v-icon>mdi-pen</v-icon></v-btn>
                <v-speed-dial
                  v-model="dial"
                  top
                  right
                  direction="bottom"
                  absolute
                  style="margin-top:-32px">
                  <template v-slot:activator>
                    <v-btn
                      v-model="dial"
                      color="primary"
                      small
                      top
                      right
                      fab
                      elevation="2">
                      <v-icon v-if="dial">mdi-close</v-icon>
                      <v-icon v-else>mdi-menu-down</v-icon>
                    </v-btn>
                  </template>
                  <v-btn
                    fab
                    dark
                    small
                    color="secondary"
                    @click="copyUrl"
                    >
                    <v-icon>mdi-link</v-icon>
                  </v-btn>
                  <v-btn
                    fab
                    color="secondary"
                    small
                    @click="showHistory">
                    <v-icon>mdi-file-eye</v-icon>
                  </v-btn>
                </v-speed-dial>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col cols="12" md="4"><SideBar/></v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, ref } from '@vue/composition-api'
import { usePage } from '@/lib/usePage'
import Loading from '@/components/Loading.vue'
import TabTitle from '@/components/app/TabTitle.vue'
import SideBar from '@/components/site/SideBar.vue'
import WikiContent from '@/components/page/WikiContent.vue'
import { useSite } from '@/lib/useSite'
import Skaldmd from '@/lib/skaldmd'
import { useProfile } from '@/lib/useProfile'
import { useAppState } from '@/lib/useAppState'
import router from '@/router'
export default defineComponent({
  components: {
    Loading,
    TabTitle,
    SideBar,
    WikiContent
  },
  setup () {
    const { site } = useSite()
    const { meta, page } = usePage()
    const { pushSnack } = useAppState()
    const dial = ref(false)
    const content = computed(() => {
      if (!page) return ''
      else if (page.value.htmlContent) return page.value.htmlContent
      const skaldmd = new Skaldmd(site.value.siteid)
      return skaldmd.toHtml(page.value.content)
    })
    const { activeProfile } = useProfile()
    const editorActions = computed(() =>
      activeProfile.value &&
      activeProfile.value.owns &&
      activeProfile.value.owns.includes(site.value.siteid))

    const copyUrl = () => {
      const dummy = document.createElement('input')
      document.body.appendChild(dummy)
      dummy.value = window.location.href
      dummy.select()
      document.execCommand('copy')
      document.body.removeChild(dummy)
      pushSnack('Success', 'Url copied to clipboard')
    }
    const showHistory = () => {
      router.push('/i/' + site.value.siteid + '/' + page.value.pageid)
    }
    return { meta, page, site, content, editorActions, dial, copyUrl, showHistory }
  }
})
</script>
