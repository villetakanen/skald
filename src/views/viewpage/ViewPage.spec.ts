import { shallowMount } from '@vue/test-utils'
import ViewPage from './ViewPage.vue'

describe('Mounted App', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(ViewPage, {
      mocks: {
        $t: () => {}
      },
      propsData: {
        siteid: 'mekanismi',
        pageid: 'ohjeita'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
