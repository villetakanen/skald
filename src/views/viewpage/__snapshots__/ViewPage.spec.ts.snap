// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted App renders correctly 1`] = `
<div
  class="page-container"
  pageid="ohjeita"
  siteid="mekanismi"
>
  <v-container-stub
    tag="div"
  >
    <!---->
     
    <v-row-stub
      tag="div"
    >
      <v-col-stub
        tag="div"
      >
        <tabtitle-stub
          link="/v//"
          sub=""
          sublink="/v/"
          topic=""
        />
      </v-col-stub>
    </v-row-stub>
     
    <v-row-stub
      tag="div"
    >
      <v-col-stub
        cols="12"
        md="8"
        tag="div"
      >
        <v-card-stub
          loaderheight="4"
          tag="div"
        >
          <v-card-text-stub>
            <div
              class="wikipage"
            >
              <div
                class=""
              >
                <wikicontent-stub
                  html=""
                />
              </div>
            </div>
             
            <v-btn-stub
              absolute="true"
              activeclass=""
              color="primary"
              disabled="true"
              fab="true"
              right="true"
              small="true"
              style="margin-right: 48px; margin-top: 4px;"
              tag="button"
              to="/edit//"
              top="true"
              type="button"
            >
              <v-icon-stub>
                mdi-pen
              </v-icon-stub>
            </v-btn-stub>
             
            <v-speed-dial-stub
              absolute="true"
              direction="bottom"
              right="true"
              style="margin-top: -32px;"
              top="true"
              transition="scale-transition"
            >
               
              <v-btn-stub
                activeclass=""
                color="secondary"
                dark="true"
                fab="true"
                small="true"
                tag="button"
                type="button"
              >
                <v-icon-stub>
                  mdi-link
                </v-icon-stub>
              </v-btn-stub>
               
              <v-btn-stub
                activeclass=""
                color="secondary"
                fab="true"
                small="true"
                tag="button"
                type="button"
              >
                <v-icon-stub>
                  mdi-file-eye
                </v-icon-stub>
              </v-btn-stub>
            </v-speed-dial-stub>
          </v-card-text-stub>
        </v-card-stub>
      </v-col-stub>
       
      <v-col-stub
        cols="12"
        md="4"
        tag="div"
      >
        <sidebar-stub />
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;
